File: lintcode/python/0628_maximum_subtree.py
#!/usr/bin/python -t

# dfs and divid and conquer

"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""

class Solution:
    def __init__(self):
        self.max_val = -sys.maxint
        self.ret_node = None
    """
    @param root: the root of binary tree
    @return: the maximum weight node
    """
    def findSubtree(self, root):
        # write your code here
        if not root:
            return root
            
        self.dfs(root)
        
        return self.ret_node
        
    def dfs(self, node):
        if not node:
            return 0
        
        l = self.dfs(node.left)
        r = self.dfs(node.right)
        
        total = node.val + l + r
        if total > self.max_val:
            self.max_val = total
            self.ret_node = node
            
        return total
        

File: lintcode/golang/0628_maximum_subtree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

/**
 * @param root: the root of binary tree
 * @return: the maximum weight node
 */
 

const MaxUint = ^uint(0)
const MinUint = 0

const MaxInt = int(^uint(0) >> 1)
const MinInt = -MaxInt - 1


func findSubtree (root *TreeNode) *TreeNode {
    // write your code here
    if root == nil {
        return nil
    }
    
    var ret TreeNode
    retNodeP := &ret
    maxVal := MinInt
    
    helper(root, &maxVal, &retNodeP)
    
    return retNodeP
}

func helper(node *TreeNode, maxVal *int, ret **TreeNode) int {
    if node == nil {
        return 0
    }
    
    left := helper(node.Left, maxVal, ret)
    right := helper(node.Right, maxVal, ret)
    
    nodeVal := left+right+node.Val
    
    if *ret == nil || nodeVal > *maxVal {
        *maxVal = node.Val
        *ret = node
    }
    
    return nodeVal
}
